crates/circuit/src/circuit_data.rs

```diff
@@ -327,6 +317,25 @@ impl CircuitData {
         Ok(res)
     }
 
+    /// Performs an empty-like shallow copy.
+    ///
+    /// Returns:
+    ///     CircuitData: The shallow copy.
+    #[pyo3(signature = (reserve = None,))]
+    pub fn copy_empty_like(&self, py: Python<'_>, reserve: Option<usize>) -> PyResult<Self> {
+        let mut res = CircuitData::new(
+            py,
+            Some(self.qubits.cached().bind(py)),
+            Some(self.clbits.cached().bind(py)),
+            None,
+            reserve.unwrap_or_default(),
+            self.global_phase.clone(),
+        )?;
+        res.qargs_interner = self.qargs_interner.clone();
+        res.cargs_interner = self.cargs_interner.clone();
+
+        Ok(res)
+    }
```

Comment:
A couple of notes:

- Since it's empty of data, there's not really a distinction between a "shallow" and "deep" copy
- As the return value is an owned value of `Self`, the `reserve` argument might be better as a separate function on `CircuitData` (if it isn't already).  It's not really any more expensive to do the `reserve` call after the creation of the internal vec.

Reply:
Addressed in dc555c8